### PR TITLE
add support for fill value in resampling tools

### PR DESCRIPTION
**Pull request by pbellec**
The nilearn resampling tools use a value of 0 for extrapolation, and it's not possible to change this behaviour.

Exposing this "fill value" as a parameter is fairly common in resampling tools, and I incidentally need it for PR #1766 

I have added a "fill" argument that maps to the `cval` argument of `scipy.ndimage.affine_transform`. This argument now appears in `_resample_one_img`, `resample_img`, and `resample_to_img`. 
The argument is optional and the default value is 0, so the new feature should be retro-compatible.

I've done some minimal testing and it seems to work.
I'll implement tests if you think it is indeed a good idea to add this feature.

### CODE REVIEW

File: nilearn/image/resampling.py [hunk 1]
```diff
@@ -287,7 +288,7 @@ def _resample_one_img(data, A, b, target_shape,
 
 def resample_img(img, target_affine=None, target_shape=None,
                  interpolation='continuous', copy=True, order="F",
-                 clip=True):
+                 clip=True, fill=0):
```
Comment (GaelVaroquaux) [inline nilearn/image/resampling.py]: I would prefer this parameter to be called "fill_value"

Comment (pbellec) [reply nilearn/image/resampling.py]: sure.

**Comment by pbellec** [comment]
will work on tests.

**Comment by pbellec** [comment]
Test added, name changed, let me know what you think.

File: nilearn/image/resampling.py [hunk 2]
```diff
@@ -118,7 +118,7 @@ def coord_transform(x, y, z, affine):
     --------
     Transform data from coordinates to brain space. The "affine" matrix
     can be found as the ".affine" attribute of a nifti image, or using
-    the "get_affine()" method for older nibabel installations::
+    the "get_affine()" method for older nibabel installations:
```
Comment (GaelVaroquaux) [inline nilearn/image/resampling.py]: No: that was correct. "::" is a special marker in restructured text that starts a verbatim block (or code block).

File: nilearn/image/tests/test_resampling.py [hunk 3]
```diff
@@ -145,6 +145,36 @@ def test_downsample():
...
+
+    data_3d = prng.rand(1, 4, 4)
+    data_4d = prng.rand(1, 4, 4, 3)
+
+    angle = np.pi/4
+    rot = rotation(0, angle)
+
+    # Try a few different fill values
+    for data in [data_3d, data_4d]:
+        for val in (-3, -1.7537, 0, 0.666, 11):
```
Comment (GaelVaroquaux) [inline nilearn/image/tests/test_resampling.py]: In the interest of test running time, could we make this list smaller, for instance keeping (-3, 0.666, 11).

Comment (GaelVaroquaux) [reply nilearn/image/tests/test_resampling.py]: Actually, I am not sure that I see the point of testing with more than 2 values.

Comment (jeromedockes) [reply nilearn/image/tests/test_resampling.py]: You could do two checks, one explicitly specifying a non-zero value, and one leaving the default and checking you get 0

File: nilearn/image/resampling.py (hunk 2 above)
Comment (pbellec) [reply nilearn/image/resampling.py]: thanks for the explanation! Reverted back.

File: nilearn/image/tests/test_resampling.py (hunk 3 above)
Comment (pbellec) [reply nilearn/image/tests/test_resampling.py]: ok, so, I went with large negative (<-1) and large positive (>1)  values with decimals, which test that clipping is indeed not performed on the filling value and possibly rounding / type errors. I also added a test of the default value (0) as suggested.

File: nilearn/image/tests/test_resampling.py [hunk 4]
```diff
@@ -145,6 +145,42 @@ def test_downsample():
...
+                rot_img = resample_img(Nifti1Image(data, np.eye(4)),
+                                       target_affine=rot,
+                                       interpolation='nearest',
+                                       fill_value=val,
+                                       clip = False)
+            else:
+                rot_img = resample_img(Nifti1Image(data, np.eye(4)),
+                                       target_affine=rot,
+                                       interpolation='nearest',
+                                       clip = False)
```
Comment (GaelVaroquaux) [inline nilearn/image/tests/test_resampling.py]:
```suggestion
                                       clip=False)
```

PEP8

File: nilearn/image/tests/test_resampling.py [hunk 5]
```diff
@@ -145,6 +145,42 @@ def test_downsample():
...
+
+    # Try a few different fill values
+    for data in [data_3d, data_4d]:
+        for val in (-3.75, 2.24, 0):
+            if val:
+                rot_img = resample_img(Nifti1Image(data, np.eye(4)),
+                                       target_affine=rot,
+                                       interpolation='nearest',
+                                       fill_value=val,
+                                       clip = False)
```
Comment (GaelVaroquaux) [inline nilearn/image/tests/test_resampling.py]:
```suggestion
                                       clip=False)
```
PEP 8

File: nilearn/image/tests/test_resampling.py [hunk 6]
```diff
@@ -145,6 +145,42 @@ def test_downsample():
...
+
+    data_3d = prng.rand(1, 4, 4)
+    data_4d = prng.rand(1, 4, 4, 3)
+
+    angle = np.pi/4
+    rot = rotation(0, angle)
+
+    # Try a few different fill values
+    for data in [data_3d, data_4d]:
+        for val in (-3.75, 2.24, 0):
```
Comment (GaelVaroquaux) [inline nilearn/image/tests/test_resampling.py]:
```suggestion
        for val in (-3.75, 0):
```
With 3D and 4D data, and three values, that's 6 calls in the test. I would like to keep it down: I am not sure what different code paths it tests, and it increases computational cost.

**Comment by GaelVaroquaux** [comment]
I'm made 3 suggested changes. You just have to click on them on github to accept them. Once this is done and the CI passes, we merge.

File: nilearn/image/tests/test_resampling.py (hunk 6 above)
Comment (jeromedockes) [reply nilearn/image/tests/test_resampling.py]: I would have like to have

- one test specifying a value, e.g. -3.75 here
- one test that doesn't specify a value, and checks we get 0, i.e. default behavior remains the same

**Comment by GaelVaroquaux** [comment]
Hey @pbellec : this PR should really just require a couple of clicks on the github interface from your side.

**Comment by pbellec** [comment]
> Hey @pbellec : this PR should really just require a couple of clicks on the github interface from your side.

Sorry, I had very little time over the past couple of days. Just pushed the changes, hopefully the tests will run ok.

**Comment by GaelVaroquaux** [comment]
Circle had timed out. I restarted it.

**Comment by GaelVaroquaux** [comment]
CI ran fine. I am merging this. Thank you!

**Comment by pbellec** [comment]
Awesome! Thanks for merging.

**Comment by kchawla-pi** [comment]
@pbellec Thanks for your excellent contribution!

**Pull request merged**